### ISSUE TITLE
ssh_config File Is Missing Nodes

### ISSUE THREAD

**Issue by mikeln** [samsung-cnct/k2#722]
Not sure when this started, but after the etcd split out.

The generated ssh_config file for the cluster is missing entries.

It seems to always have the entries for both etcd clusters (etcd, etcdEvents).
But sometimes has nothing else, sometimes has one, or a few, of the other entries:
master, specialNodes, clusterNodes.

**Comment by alejandroEsc** [comment]
I can confirm this. I suspect order in which things happen.

**Comment by mikeln** [comment]
Not consistently failing..   I sometimes get the full file.

**Comment by mikeln** [comment]
When the cluster start fails...on a "re-up"... the values in the ssh_config are not updated (or maybe I should say overwritten).   Makes it very hard to debug problems when, for example, the master-n  IP is incorrect.

**Comment by mikeln** [comment]
Just had a full cluster start with no errors... however, the ssh_config has entries for all the nodes, but some of them (notably the master-x) have the wrong IP addresses.

**Comment by alejandroEsc** [comment]
@mikeln could you print out that ssh_config file.

**Comment by mikeln** [comment]
My cluster: 5,5,3,3,2.    master-3 is missing, all non-etcd/events nodes have incorrect IPs:
```
mikeln-bender:.kraken Mikel_Nelson$ cat mikeln-v17/ssh_config
Host etcd-1
  HostName ec2-54-201-16-179.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
Host etcd-2
  HostName ec2-54-186-249-150.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
Host etcd-3
  HostName ec2-54-218-123-136.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
Host etcd-4
  HostName ec2-54-190-4-86.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
Host etcd-5
  HostName ec2-34-210-86-199.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
Host etcdEvents-1
  HostName ec2-54-218-201-109.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
Host etcdEvents-2
  HostName ec2-54-213-5-85.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
Host etcdEvents-3
  HostName ec2-34-223-250-33.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
Host etcdEvents-4
  HostName ec2-34-209-244-47.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
Host etcdEvents-5
  HostName ec2-54-186-87-103.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
Host master-1
  HostName ec2-54-191-61-69.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
Host master-2
  HostName ec2-54-213-240-167.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
Host clusterNodes-1
  HostName ec2-34-211-112-62.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
Host clusterNodes-2
  HostName ec2-54-202-140-27.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
Host clusterNodes-3
  HostName ec2-34-211-226-172.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
Host specialNodes-1
  HostName ec2-34-213-181-45.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
Host specialNodes-2
  HostName ec2-54-245-152-173.us-west-2.compute.amazonaws.com
  Port 22
  User core
  IdentityFile /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.ssh/mln_id_rsa
```

**Comment by alejandroEsc** [comment]
is this for the cluster mikeln-v17?

**Comment by mikeln** [comment]
yes

**Comment by alejandroEsc** [comment]
@mikeln I only really see master-nodes be a problem. Also i would not trust this config file with things managed by an ASG, specially in the bootstrap process. My suggestions is to remove the etcd log and rerurn up with --tags ssh to renew and verify once again.

**Comment by mikeln** [comment]
ok, but I'm not looking for a way to fix my file.   

The issue is that this wasn't created correctly in the first place, and this problem seems to have just reappeared.

master-1,master-2,clusterNode-1,clusterNode-2,clusterNode-3 are wrong.  master-3 is missing. specialNode-1,2 are correct.

**Comment by alejandroEsc** [comment]
I see each cluster node (workers) have the correct public ips, the -1 numbers are arbitrary, am I missing something? Also yes i agree there are items missing but I cant tell if this is a new problem or just always there, and my best guess is that ssh_config is probably not placed in an ideal location in sequence? I dont know.

**Comment by mikeln** [comment]
yes, you are correct about clusterNodes, I was looking at the wrong EC2 instances for those (refresh helps :) ).

**Comment by alejandroEsc** [comment]
Yup, just making sure my changes didnt further break this, doesnt seem to be the case.

So my case about bootstrapping and asgs being the issue here:

I have consistently seen etcd nodes labeled correctly, Not surprised since these are very high up in the sequence, are not managed by asg and usually are not killed/die in the process. 

I have seen all other nodes either missing one item, or all items for that group. These tend to be managed by asg and usually die in bootstrapping, especially when etcd nodes are having trouble with quorum and in extreme cases when you have a split brain scenario. I would also mention that these are brought up further down wind from etcd being up.

**Comment by samba** [comment]
Possibly related? #742

**Comment by samba** [comment]
I've looked into this in some depth now, and have a few fixes to the general `ssh_config` population and state interpolation from AWS API. This addresses inconsistencies of IP addresses within the non-ASG nodepools, where names/aliases can be static.

@mikeln In the event of cluster setup failure, I think our best case is to re-create the SSH config with the nodes that are actually available. I've confirmed that @alejandroEsc's suggestion of running with `--tags` can provide this, and tuned it a bit too... PR pending.
